### PR TITLE
init

### PR DESCRIPTION
**Pull request by hexfusion**
Signed-off-by: Sam Batschelet <sbatsche@redhat.com>

Please read https://github.com/etcd-io/etcd/blob/master/CONTRIBUTING.md#contribution-flow.

### CODE REVIEW

File: pkg/transport/listener.go [hunk 1]
```diff
@@ -63,6 +66,48 @@ func wrapTLS(scheme string, tlsinfo *TLSInfo, l net.Listener) (net.Listener, err
...
+}
+
+func newListenConfig(addr, scheme string, sopts SocketOpts)  (net.ListenConfig, error) {
+	lc := net.ListenConfig{}
+	if len(sopts) > 0 {
+		lc.Control = sopts.Control
+	}
+	return lc, nil
+}
+
```
Comment (hexfusion) [inline pkg/transport/listener.go:93]: move to socketopt_linux.go

File: pkg/transport/listener.go [hunk 2]
```diff
@@ -63,6 +66,48 @@ func wrapTLS(scheme string, tlsinfo *TLSInfo, l net.Listener) (net.Listener, err
 	return newTLSListener(l, tlsinfo, checkSAN)
 }
 
+// NewListenerWithConfig
```
Comment (hexfusion) [inline pkg/transport/listener.go:69]: add details

File: pkg/transport/listener.go [hunk 3]
```diff
@@ -63,6 +66,48 @@ func wrapTLS(scheme string, tlsinfo *TLSInfo, l net.Listener) (net.Listener, err
...
+func NewListenerWithConfig(addr, scheme string, tlsinfo *TLSInfo, sopts SocketOpts) (net.Listener, error) {
```
Comment (hexfusion) [inline pkg/transport/listener.go:70]: NewListenerWithSocketOpts

File: pkg/transport/listener.go [hunk 4]
```diff
@@ -63,6 +66,48 @@ func wrapTLS(scheme string, tlsinfo *TLSInfo, l net.Listener) (net.Listener, err
...
+func (sopts SocketOpts) Control(network, addr string, conn syscall.RawConn) error {
+	for _, s := range sopts {
+		if err := s(network, addr, conn); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func SetReuseAddress(network, address string, conn syscall.RawConn) error {
```
Comment (hexfusion) [inline pkg/transport/listener.go:105]: no export

File: pkg/transport/timeout_listener.go [hunk 5]
```diff
@@ -27,15 +29,28 @@ func NewTimeoutListener(addr string, scheme string, tlsinfo *TLSInfo, rdtimeoutd
 	if err != nil {
 		return nil, err
 	}
-	ln = &rwTimeoutListener{
+	return newTimeoutListener(ln, addr, scheme, rdtimeoutd, wtimeoutd, tlsinfo)
+}
+
+func NewTimeoutListerWithConfig(addr string, scheme string, tlsinfo *TLSInfo, rdtimeoutd, wtimeoutd time.Duration, sopts SocketOpts ) (net.Listener, error) {
```
Comment (hexfusion) [inline pkg/transport/timeout_listener.go:35]: WithSocketOpts

File: pkg/transport/transport.go [hunk 6]
```diff
@@ -15,6 +15,7 @@
 package transport
 
 import (
+
```
Comment (hexfusion) [inline pkg/transport/transport.go:18]: drop

File: pkg/transport/transport.go [hunk 7]
```diff
@@ -69,3 +70,4 @@ func (urt *unixTransport) RoundTrip(req *http.Request) (*http.Response, error) {
 	req.URL.Scheme = strings.Replace(req.URL.Scheme, "unix", "http", 1)
 	return urt.Transport.RoundTrip(req)
 }
+
```
Comment (hexfusion) [inline pkg/transport/transport.go:73]: drop

File: pkg/transport/unix_listener.go [hunk 8]
```diff
@@ -15,8 +15,11 @@
 package transport
```
Comment (hexfusion) [inline pkg/transport/unix_listener.go:15]: reset

File: server/embed/config.go [hunk 9]
```diff
@@ -241,6 +241,8 @@ type Config struct {
 
 	CORS map[string]struct{}
 
+
+
```
Comment (hexfusion) [inline server/embed/config.go:245]: drop

File: server/embed/config.go [hunk 10]
```diff
@@ -306,6 +308,15 @@ type Config struct {
...
 	ListenMetricsUrlsJSON string `json:"listen-metrics-urls"`
 
+	// ReuseAddress enables a socket option SO_REUSEADDR[1] which informs the
+	// kernel to tolerate a busy port in `TIME_WAIT` state. This allows resuse
+	// of local addresses. Useful in cases where etcd process has been
+	// restarted, without this option active sockets on that port will block
+	// the new process from listening. Depending on the situtation this wait
+	// time can be considerable (many seconds) and degrade MTTR.
+	// [1] https://man7.org/linux/man-pages/man7/socket.7.html
+	ReuseAddress bool `json:"reuse-address"`
```
Comment (hexfusion) [inline server/embed/config.go:318]: naming are we reuse or sharing

File: server/embed/config.go [hunk 11]
```diff
@@ -551,7 +563,7 @@ func updateCipherSuites(tls *transport.TLSInfo, ss []string) error {
 	}
 	return nil
 }
-
+re
```
Comment (hexfusion) [inline server/embed/config.go:566]: drop

File: server/embed/etcd.go [hunk 12]
```diff
@@ -523,6 +524,14 @@ func (e *Etcd) servePeers() (err error) {
 	return nil
 }
 
+func getSocketOptions(cfg *Config) transport.SocketOpts {
```
Comment (hexfusion) [inline server/embed/etcd.go:527]: move to socket file and export

File: server/embed/etcd.go [hunk 13]
```diff
@@ -565,7 +574,7 @@ func configureClientListeners(cfg *Config) (sctxs map[string]*serveCtx, err erro
 			continue
 		}
 
-		if sctx.l, err = net.Listen(network, addr); err != nil {
+		if sctx.l, err = transport.NewListenerWithConfig(addr, u.Scheme, &cfg.ClientTLSInfo, getSocketOptions(cfg)); err != nil {
 			return nil, err
```
Comment (hexfusion) [inline server/embed/etcd.go:578]: WithSocketOpts

File: server/embed/etcd.go [hunk 14]
```diff
@@ -658,6 +667,7 @@ func (e *Etcd) serveClients() (err error) {
 	// start client servers in each goroutine
 	for _, sctx := range e.sctxs {
 		go func(s *serveCtx) {
+			// here
```
Comment (hexfusion) [inline server/embed/etcd.go:670]: drop

File: server/etcdserver/api/rafthttp/util.go [hunk 15]
```diff
@@ -42,6 +42,10 @@ func NewListener(u url.URL, tlsinfo *transport.TLSInfo) (net.Listener, error) {
 	return transport.NewTimeoutListener(u.Host, u.Scheme, tlsinfo, ConnReadTimeout, ConnWriteTimeout)
 }
 
+func NewListenerWithSocketOptions(u url.URL, tlsinfo *transport.TLSInfo, ctls transport.SocketOpts ) (net.Listener, error) {
+	return transport.NewTimeoutListerWithConfig(u.Host, u.Scheme, tlsinfo, ConnReadTimeout, ConnWriteTimeout, ctls)
```
Comment (hexfusion) [inline server/etcdserver/api/rafthttp/util.go:46]: export?

File: pkg/transport/listener.go [hunk 16]
```diff
@@ -29,7 +30,9 @@ import (
 	"os"
 	"path/filepath"
 	"strings"
+	"syscall"
 	"time"
+	"golang.org/x/sys/unix"
```
Comment (hexfusion) [inline pkg/transport/listener.go:35]: no core

**Comment by ptabor** [comment]
`./scripts/fix.sh` usually helps with refreshing all go.sum files. 
or manual 'go mod tidy'

Please describe what the PR is accomplishing in the description.

**Comment by hexfusion** [comment]
excuse the noise was mean for local repo

**Pull request closed without merging**